### DISCUSSION TITLE
Search buffers and files at the same time

### DISCUSSION THREAD

**Discussion by jmarquis** [ibhagwan/fzf-lua#1483] [General]
I've recently been trying out LazyVim and figured I'd give Telescope another shot as part of that since it's LazyVim's default picker, and I installed [smart-open.nvim](https://github.com/danielfalk/smart-open.nvim) in an attempt to give me that "magic" ordering that I've come to love from fzf-lua (and previously fzf.vim). Long story short, smart-open is pretty cool but ultimately it didn't scratch the itch the same way that fzf-lua does for me, for various reasons. (Also Telescope's performance with gigantic codebases still isn't even close to fzf-lua or fzf.vim all these years later, even with fd/rg/etc.) All this has really made me appreciate the thoughtfulness you and @junegunn have put into this space even more than I already did.

While I didn't care for the "smart" aspects of smart-open, one thing I came to appreciate was that it combines a bunch of file-related sources into one "open anything" list, and I particularly got used to being able to swap to open buffers _or_ search all files in my cwd using the same keybind. I used `;` for it, so I could just hit `;` and see all my open buffers at the top, followed by the usual live file results, and search all of them at once (with buffers prioritized in the results). This means if I want to jump to a file, I don't have to remember if I have it open in a buffer or not, I can just hit `;` and start typing. This is particularly nice in massive codebases where the `fd` results can take a few seconds to come back, so I can still search my open buffers very quickly.

Anyway I'm happily back on fzf-lua, but now I have muscle memory to jump to any buffer or file using `;`. So, here's my initial attempt at recreating this functionality. I haven't tested it a ton yet but it seems to be working in a few smaller directories so far. I'm curious if anyone has suggestions for improvements, particularly any way to take any advantage of the more nuanced `buffers` logic built in to fzf-lua. (Or you know, if @ibhagwan wants to ship a built in buffers_and_files picker that does something like this.... ;) )

```lua
local cmd = ""
local buffers = {}
for _, bufnr in ipairs(vim.api.nvim_list_bufs()) do
  if vim.api.nvim_buf_is_loaded(bufnr) then
    local bufname = vim.api.nvim_buf_get_name(bufnr)
    if bufname ~= "" then
      bufname = vim.fn.fnamemodify(bufname, ":~:.")
      if
        string.find(vim.fs.basename(bufname), "NvimTree_") ~= 1 -- filter out nvim tree buffer
        and bufname ~= vim.fn.fnamemodify(vim.fn.expand("%"), ":~:.") -- filter out current buffer
      then
        local info = vim.fn.getbufinfo(bufnr)
        table.insert(buffers, { name = bufname, info = info[1] or info })
      end
    end
  end
end

table.sort(buffers, function(a, b)
  return a.info.lastused > b.info.lastused
end)

for _, buffer in ipairs(buffers) do
  cmd = cmd .. 'echo "' .. buffer.name .. '" && '
end

cmd = cmd .. "fd --color=never --type f --hidden --follow --exclude .git"

for _, buffer in ipairs(buffers) do
  cmd = cmd .. ' --exclude "' .. buffer.name .. '"'
end

require("fzf-lua").files({ cmd = cmd, fzf_opts = { ["--tiebreak"] = "index" } })
```

**Comment by ibhagwan** [comment]
Ty for the kind words @jmarquis.

I like the approach you took combining the results into a shell command and into `files`, this way you can still keep the performance of `fd` and fzf as an external processes freeing the main neovim thread.

> Or you know, if @ibhagwan wants to ship a built in buffers_and_files picker that does something like this.... ;)

Not yet, as this would require supporting all the buffer flags, options, sorting, etc as well as a windows adaptation of the command (may or may not be annoying, such as Windows :)).

Your solution is wonderful, with a bit of effort you can make the display equal to the fzf-lua buffer entries with flags, buffer numbers and ansi colors, since `path.entry_to_file` can handle both buffers and files everything else would be seamless (actions / previewers) - if you do decide to embark on this road I’ll be happy to help point out where you can reuse certain codes.

**Reply by jmarquis** [reply]
I probably will iterate in that direction! I'll check out `path.entry_to_file`, somehow I didn't stumble across this when I was reading through the `files` and `buffers` code. I'm going to see how it feels in that big codebase I mentioned when I go back to work tomorrow, and see what seems to be missing.

**Comment by ibhagwan** [comment]
> I used ; for it,

Btw, IMHO you’re  missing out on a very important motion, in normal mode, once you find a character with `f<char>`, you can then repeat the same search with `;` forwards and `,` backwards. I find this very useful when navigating the line as sometime I would press `f”`, for example but I would hit an earlier `”` that I missed, instead of pressing `f”` again I would press `;`.

**Reply by jmarquis** [reply]
Yeah, you're completely right 😅 I've always meant to un-learn the muscle memory for swapping buffers with `;`. I picked that key because when I first learned vim years ago I didn't understand what `f` or `;` did, and still under-use those motions today. Maybe I'll try putting `<leader>` in front of the fzf-lua keybind and see if my brain will accept it after a few days.

**Reply by ibhagwan** [reply]
> Yeah, you're completely right 😅 I've always meant to un-learn the muscle memory for swapping buffers with `;`. I picked that key because when I first learned vim years ago I didn't understand what `f` or `;` did, and still under-use those motions today. Maybe I'll try putting `<leader>` in front of the fzf-lua keybind and see if my brain will accept it after a few days.

`leader;` used to be my original buffers mapping, I now use `leader,` (leader-comma) instead which I find the most convenient on my external keyboard and laptop keyboard:
https://github.com/ibhagwan/nvim-lua/blob/830c0a88651302d43487e05b3423c72ac77ed4a2/lua/plugins/fzf-lua/mappings.lua#L50
